### DISCUSSION TITLE
E339: Pattern too long

### DISCUSSION THREAD

**Discussion by jaidetree** [hrsh7th/nvim-cmp#2104] [Q&A]
For the better part of a year now been dealing with an issue where nvim-cmp just kinda stops working. Normally it would completely crash my instance of nvim but as of a week ago I disabled autocomplete in favor of just pressing `<C-Space>`. This didn't fix the issue but now I get an error when I invoke the completion UI but nvim still runs, but I can't use nvim-cmp until I restart nvim.

### What I know

- It only happens on TypeScript projects
- LSP is enabled and is one of my sources
- Entry point is invoking the `cmp.mapping.complete` function
- I do not encounter this error on ClojureScript projects, seems to be related to TypeScript
- The error seems to build up over time. I usually get about ~2 hours after a clean restart until the error.
- Once the error first occurs, I get the error every time I invoke nvim-cmp until I restart neovim
- Switching to manual completion prevents the error from crashing neovim.
- The error occurs regardless of what I'm trying to complete against. For example, if I wrote `interna|` then trigger completion at the end, I can still encounter that error at the same time if I wrote `i|` and tried to complete from that one character. Suggests this isn't directly about the completion target.

### The Error Message

```lua
E339: Pattern too long
E5108: Error executing lua: ...local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/pattern.lua:13: regex: internal error
stack traceback:
        [C]: in function 'match_str'
        ...local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/pattern.lua:13: in function 'offset'
        ...rs/j/.local/share/nvim/lazy/nvim-cmp/lua/cmp/context.lua:77: in function 'callback'
        .../.local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/cache.lua:38: in function 'get_offset'
        /Users/j/.local/share/nvim/lazy/nvim-cmp/lua/cmp/source.lua:281: in function 'complete'
        /Users/j/.local/share/nvim/lazy/nvim-cmp/lua/cmp/core.lua:299: in function 'complete'
        /Users/j/.local/share/nvim/lazy/nvim-cmp/lua/cmp/init.lua:77: in function 'complete'
        ...ocal/share/nvim/lazy/nvim-cmp/lua/cmp/config/mapping.lua:139: in function 'on_keymap'
        /Users/j/.local/share/nvim/lazy/nvim-cmp/lua/cmp/core.lua:145: in function 'callback'
        ....local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/keymap.lua:139: in function <....local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/keymap.lua:133>

```

### My Config

I'm using fennel but my config is not much beyond what was recommended in the readme:

```fennel
(cmp.setup {:snippet {:expand expand-snippet}
            :completion {:autocomplete false}
            :mapping {:<C-Space> (cmp.mapping.complete)
                      :<CR> (cmp.mapping.confirm {:select false})
                      :<Tab> (cmp.mapping.select_next_item)
                      :<S-Tab> (cmp.mapping
                                 (fn [fallback]
                                   (if 
                                     (cmp.visible) (cmp.select_prev_item)
                                     (luasnip.jumpable -1) (luasnip.jump -1)
                                     (fallback))))
                      :<C-b> (cmp.mapping.scroll_docs -4)
                      :<C-f> (cmp.mapping.scroll_docs 4)
                      :<C-g> (cmp.mapping.abort)}
            :sources (cmp.config.sources [{:name :nvim_lsp_signature_help}
                                          {:name :nvim_lsp}]
                                         [{:name :luasnip}]
                                         [{:name :path}
                                          {:name :cmp_git}]
                                         [{:name :buffer}])})

(cmp.setup.filetype :gitcommit
                    {:sources (cmp.config.sources [{:name :cmp_git}
                                                   {:name :buffer}])})

(cmp.setup.cmdline "/" {:mapping (cmp.mapping.preset.cmdline)
                        :sources [{:name :buffer}]})

(cmp.setup.cmdline ":"
                   {:mapping (cmp.mapping.preset.cmdline)
                    :sources (cmp.config.sources [{:name :path}
                                                  {:name :cmdline}])})
```

Plus some LSP config:

```fennel
(local capabilities (vim.tbl_deep_extend
                      :force
                      (vim.lsp.protocol.make_client_capabilities)
                      (cmp-nvim-lsp.default_capabilities)
                      {:workspace {:didChangeWatchedFiles {:dynamicRegistration true}}}))
```

### Why here?

I only encounter this error after about 2 hours within a nvim session so I'm not sure I can produce a minimal reproduction repo as I'm still trying to learn how to reproduce it immediately. I am trying to dive in, discover the cause, and fix it but at this point I feel stuck so looking for advice on where to look. Also curious if anyone is running into this error too.

**Comment by Shougo** [comment]
Hm.
the error means the string pattern is too long.

**Reply by jaidetree** [reply]
That part I get 😁 but not sure why that would happen. This time I only invoked the completion UI at the end of a word. What pattern do you think this refers to? The input I'm trying to complete?

**Reply by Shougo** [reply]
I think the input, I have checked the code and I don't know why the input is too long.

**Reply by jaidetree** [reply]
Right? I poked around too and it didn't jump out at me what would cause that. Any hints for how I could log the input pattern it's testing against? Then when it crashes next time the input should be right above it.

**Reply by jaidetree** [reply]
https://github.com/hrsh7th/nvim-cmp/blob/ca4d3330d386e76967e53b85953c170658255ecb/lua/cmp/utils/pattern.lua#L13

I think I can monkey patch that function with something like:

```fennel
(local pattern-utils (require :cmp.utils.patterns))

(local original-offset-function pattern-utils.offset)

(tset pattern-utils 
      :offset (fn [p text] 
                  (message (.. "cmp.utils.pattern " p " " text))
                  (original-offset-function p text)))
```

**Reply by jaidetree** [reply]
Have not hit that issue yet but I did type `expor` in a TypeScript file and invoked completion.

I did see some unexpected input:

```
cmp.utils.pattern \%(\%(\V(\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V,\m\|\V<\m\|\V)\m\|\V \m\)\s*\zs\)\m$
cmp.utils.pattern ^\%(\%(\V(\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V,\m\|\V<\m\|\V)\m\|\V \m\)\s*\zs\)
cmp.utils.pattern \%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)\m$
cmp.utils.pattern ^\%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)
cmp.utils.pattern ^\%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)
cmp.utils.pattern ^\%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)
cmp.utils.pattern \%(\%([^/\\:\*?<>'"`\|]\)*\)\m$
cmp.utils.pattern ^\%(\%([^/\\:\*?<>'"`\|]\)*\)
cmp.utils.pattern \%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\%(-\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\)*\)\)\m$
cmp.utils.pattern ^\%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\%(-\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\)*\)\)
cmp.utils.pattern \%(\%(\V(\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V)\m\|\V,\m\|\V<\m\|\V)\m\|\V \m\)\s*\zs\)\m$
cmp.utils.pattern \%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)\m$
cmp.utils.pattern \%(\%([^/\\:\*?<>'"`\|]\)*\)\m$
cmp.utils.pattern \%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\%(-\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\)*\)\)\m$


cmp.utils.pattern \%(\%([^/\\:\*?<>'"`\|]\)*\)\m$
cmp.utils.pattern ^\%(\%([^/\\:\*?<>'"`\|]\)*\)
cmp.utils.pattern \%([^[:blank:]]*\)\m$
cmp.utils.pattern ^\%([^[:blank:]]*\)
```

My guess is that is not what the `p` argument is supposed to look like? Some of those are quite long already, so maybe something is growing and it's just maxing out?

**Reply by Shougo** [reply]
I think you can catch the error and print the input and pattern.

Hm...  Your pattern is too long.
I think it is from LSP side.

**Reply by Shougo** [reply]
The pattern is from source config or LSP client specific options.

```
## Option

`[%LSPCONFIG-NAME%].keyword_pattern`

You can override keyword_pattern for specific language-server like this.

```lua
cmp.setup {
  ...
  sources = {
    {
      name = 'nvim_lsp',
      option = {
        php = {
          keyword_pattern = [=[[\%(\$\k*\)\|\k\+]]=],
        }
      }
    }
  }
  ...
}
```

**Reply by jaidetree** [reply]
Thanks for the insight! I'll look into optimizing that, but at least I have a direction now. Thanks again 🙏

**Reply by jaidetree** [reply]
https://gist.github.com/jaidetree/2fdf5f6efdc84e51f138518a04050a16

Hit that issue but this time I had my logging in place. It seems like that pattern keeps growing and growing until it just breaks. I wonder if I've messed something up where something keeps trying to setup the lsp source but instead of replacing the patterns it just keeps appending to it.

This is likely outside the scope of nvim-cmp though as you said it's one of the LSP client options but I think posting at least in this discussion makes sense as if anyone encounters this in the future the stack trace identifies nvim-cmp.

Not 100% sure where to go from here but I think the next step is finding where that pattern is coming from.

**Comment by jaidetree** [comment]
Made some progress. The culprit seems to be https://github.com/hrsh7th/cmp-nvim-lsp-signature-help, but not sure why yet.

With that plugin commented out on a fresh install I see:

```lua
cmp.utils.pattern \%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)\m$
cmp.utils.pattern ^\%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)
cmp.utils.pattern ^\%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)
cmp.utils.pattern ^\%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\w*\%(-\w*\)*\)\)
cmp.utils.pattern \%(\%([^/\\:\*?<>'"`\|]\)*\)\m$
cmp.utils.pattern ^\%(\%([^/\\:\*?<>'"`\|]\)*\)
cmp.utils.pattern \%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\%(-\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\)*\)\)\m$
cmp.utils.pattern ^\%(\%(-\?\d\+\%(\.\d\+\)\?\|\h\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\%(-\%(\w\|á\|Á\|é\|É\|í\|Í\|ó\|Ó\|ú\|Ú\)*\)*\)\)
```

The logged patterns seem static and don't change, with cmp-nvim-lsp-signature-help enabled that's when I see the long patterns that keep growing like in the previous gist: https://gist.github.com/jaidetree/2fdf5f6efdc84e51f138518a04050a16.

This is at least a big improvement, but not sure why this happens on TypeScript projects yet.
